scripts/rpmbuildinfo

```diff
@@ -0,0 +1,96 @@
+# Copyright (c) 2016  Dennis Gilmore <dennis@ausil.us>
+# Copyright (c) 2016  Simon "HW42" Geiser <hw42@ipsumj.de>
+# Copyright (c) 2016  Wojciech Porczyk <woju@invisiblethingslab.com>
+# Copyright (c) 2019  Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
+# Copyright (c) 2021  Frédéric Pierret <frederic.pierret@qubes-os.org>
+#
+# All rights reserved.
+# 
+# Permission to use, copy, modify, and distribute this software for any purpose
+# with or without fee is hereby granted, provided that the above copyright
+# notice and this permission notice appear in all copies.
+# 
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT OF THIRD PARTY RIGHTS. IN
+# NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+# OR OTHER DEALINGS IN THE SOFTWARE.
+# 
+# Except as contained in this notice, the name of a copyright holder shall not
+# be used in advertising or otherwise to promote the sale, use or other dealings
+# in this Software without prior written authorization of the copyright holder.
+#
+
+set -e
+
+RPM_BUILD_ROOT="$1"
+RPM_BUILD_DIR="$2"
+RPM_PACKAGE_EPOCH="${3:-0}"
+RPM_PACKAGE_NAME="$4"
+RPM_PACKAGE_VERSION="$5"
+RPM_PACKAGE_RELEASE="$6"
+RPM_ARCH="$7"
+
+getos() {
+    # shellcheck disable=SC1091
+    test -r /etc/os-release && . /etc/os-release
+    if test -n "$BUILD_ORIGIN"; then
+        ID="$BUILD_ORIGIN"
+    fi
+    if test -z "$ID"; then
+        ID=$(cat /etc/system-release)
+    fi
+    printf %s "$ID"
+}
+
+RPM_BUILD_NAME="${RPM_PACKAGE_NAME}-${RPM_PACKAGE_VERSION}-${RPM_PACKAGE_RELEASE}.${RPM_ARCH}"
+
+BUILDINFO_DIR="$RPM_BUILD_ROOT/usr/lib/src/buildinfo"
+BUILDINFO="${BUILDINFO_DIR}/${RPM_BUILD_NAME}.buildinfo"
+
+mkdir -p "$BUILDINFO_DIR"
+
+cat > "$BUILDINFO" <<EOF
+Format: 1.0-rpm
+Build-Architecture: $(uname -m)
+Source: $RPM_PACKAGE_NAME
+Epoch: $RPM_PACKAGE_EPOCH
+Version: ${RPM_PACKAGE_VERSION}
+Release: ${RPM_PACKAGE_RELEASE}
+Architecture: $RPM_ARCH
+Build-Origin: $(getos)
+Build-Path: $RPM_BUILD_DIR
+EOF
+
+printf 'Installed-Build-Depends:\n' >> "$BUILDINFO"
+rpm -qa --queryformat '%{epoch}:%{name}-%{version}-%{release}.%{arch}\n' \
+    | LC_ALL=C sort -t: -k2 \
+    | sed -e 's/^(none)://; /\.(none)$/d; s/^/ /' >> "$BUILDINFO"
+
+printf 'Environment:\n' >> "$BUILDINFO"
+
+# Whitelist from Debian's Dpkg:
+# https://anonscm.debian.org/git/dpkg/dpkg.git/tree/scripts/Dpkg/Build/Info.pm#n50
+ENV_WHITELIST=
+
+# Toolchain.
+ENV_WHITELIST="$ENV_WHITELIST CC CPP CXX OBJC OBJCXX PC FC M2C AS LD AR RANLIB MAKE AWK LEX YACC"
+# Toolchain flags.
+ENV_WHITELIST="$ENV_WHITELIST CFLAGS CPPFLAGS CXXFLAGS OBJCFLAGS OBJCXXFLAGS GCJFLAGS FFLAGS LDFLAGS ARFLAGS MAKEFLAGS"
```

Comment:
I don't understand your point. This is not to set with %configure. It's the job of rebuilder like https://github.com/fepitre/rpmreproduce

Reply:
@voxik @Conan-Kudo any advice on the best way to create this hook?

Reply:
I'm tempted to try something like:
```
%__buildinfo_build_post \
  env > "%{buildroot}/build.env"\
%{nil}
```

Reply:
Well, that would do no good unless you have the regular `%__build_post` include a check for it to be defined to run that step. But it's probably the way to go.

Reply:
And you'll probably want to capture the environment of the `%prep`, `%build` and `%install` stages, since mutations in the build occur in each of those stages.

Reply:
I'll try to create per build stage env file that would be triggered by corresponding %__STAGE_post

Reply:
Also don't forget `%check`.

Reply:
I may be missing something, but I think it only matters what environment variables are presented _to_ rpmbuild, not really the resulting state in each stage. That's because the actual variable values in `%build` etc should be determined by the initial environment (in which rpmbuild is started) modified with whatever rpm macros and the spec itself does.

The information here is used to reproduce the environment in which rpmbuild was called, not for manually calling individual stages of the spec outside of rpmbuild. So, inability to capture whatever `%configure` or similar does should be okay.

Reply:
Then we should be capturing at the beginning of each stage, no? We need the variables rpmbuild initializes the environment with.

Reply:
Isn't it enough to catch them once, for example in `%prep`? Should be the same, right?

Reply:
They are not. Different variables are initialized at different stages.

Reply:
Is there a way to get env variables `rpmbuild` was started in (before it modified them)?

Reply:
`/proc/self/environ`?

Reply:
The environment in various stages of the build depends on various packages in the build environment, so I don't think we should try to capture it. It shoud be reproduced when the same packages are installed. But it seems that `rpm` passes the environment passed in from the outside into the build environment, so capturing that seems useful. `/proc/self/environ` sounds like a reasonable option.

OTOH, if the builds are done in `mock`, the environment would be sanitized anyway. `mock` now uses `systemd-nspawn`, and it standarizes the environment to a large degree. Nevertheless, it does inject some information, like `container_host_*`. This was added in https://github.com/systemd/systemd/pull/15891/commits/e1bb4b0d1dd91ea13792e9d16d4bbf4ab8385f58. I wonder if we should add a switch to systemd-nspawn to opt out of setting those variables and make mock use it. The environment inside of mock should not care at all about the host.